src/_utils/focusVisible/FocusVisibleProvider.tsx

```diff
@@ -0,0 +1,110 @@
+import React, { createContext, useEffect, useState, useMemo, ReactNode } from 'react'
+
+const KEYCODE_TRIGGERING_FOCUS_VISIBLE = [
+  37, // Arrow left
+  38, // Arrow up
+  39, // Arrow right
+  30, // Arrow down
+  9, // Tab
+]
+
+// A React hook based on: https://github.com/WICG/focus-visible
+
+interface FocusVisibleProviderProps {
+  children: ReactNode
+}
+
+export const FocusVisibleContext = createContext({
+  hadKeyboardEvent: false,
+  isInitialized: false,
+})
+
+export const FocusVisibleProvider = ({ children }: FocusVisibleProviderProps) => {
+  /* When the provider first loads, assume the user is in pointer modality. */
+  const [hadKeyboardEvent, setHadKeyboardEvent] = useState(false)
+
+  useEffect(() => {
+    let lastClientX: Number
+    let lastClientY: Number
+
+    const onPointerDown = () => {
+      setHadKeyboardEvent(false)
+    }
+
+    const onInitialPointerMove = () => {
+      setHadKeyboardEvent(false)
+    }
+
+    const getLastMouseMove = (e: MouseEvent) => {
+      // Ensure the mouse has actually moved (Safari)
+      // https://transitory.technology/mouse-move-in-safari/
+      if (lastClientX === e.clientX && lastClientY === e.clientY) {
+        return
+      }
+
+      lastClientX = e.clientX
+      lastClientY = e.clientY
+    }
+
+    const addInitialPointerMoveListeners = () => {
+      document.addEventListener('mousemove', (event: MouseEvent) => {
+        getLastMouseMove(event)
+        onInitialPointerMove
+      })
+      document.addEventListener('mousedown', onInitialPointerMove)
```

Comment:
isnt possible to attach events to document dynamically? (on a `for` for instance)

Reply:
Done